src/client/datascience/jupyter/kernels/helpers.ts

```diff
@@ -282,6 +279,113 @@ export function isLocalLaunch(configuration: IConfigurationService) {
     return false;
 }
 
+export function findPreferredKernelIndex(
+    kernels: KernelConnectionMetadata[],
+    resource: Resource,
+    languages: string[],
+    notebookMetadata: nbformat.INotebookMetadata | undefined,
+    interpreter: PythonEnvironment | undefined,
+    remoteKernelPreferredProvider: PreferredRemoteKernelIdProvider | undefined
+) {
+    let index = -1;
+
+    // First try remote
+    if (index < 0 && resource && remoteKernelPreferredProvider) {
+        const preferredKernelId = remoteKernelPreferredProvider.getPreferredRemoteKernelId(resource);
+        if (preferredKernelId) {
+            // Find the kernel that matches
+            index = kernels.findIndex(
+                (k) => k.kind === 'connectToLiveKernel' && k.kernelModel.id === preferredKernelId
+            );
+        }
+    }
+
+    // If still not found, look for a match based on notebook metadata and interpreter
+    if (index < 0) {
+        const nbMetadataLanguage =
+            !notebookMetadata || isPythonNotebook(notebookMetadata)
+                ? PYTHON_LANGUAGE
+                : (
+                      (notebookMetadata?.kernelspec?.language as string) || notebookMetadata?.language_info?.name
+                  )?.toLowerCase();
+        let bestScore = -1;
+        for (let i = 0; kernels && i < kernels?.length; i = i + 1) {
+            const metadata = kernels[i];
+            const spec = metadata.kind !== 'connectToLiveKernel' ? metadata.kernelSpec : undefined;
+            const speclanguage = getKernelConnectionLanguage(metadata);
+            let score = -1;
+
+            if (spec) {
+                // See if the path matches.
+                if (
+                    spec &&
+                    spec.path &&
+                    spec.path.length > 0 &&
+                    interpreter &&
+                    spec.path === interpreter.path &&
+                    nbMetadataLanguage === PYTHON_LANGUAGE
+                ) {
+                    // Path match. This is worth more if no notebook metadata as that should
+                    // match first.
+                    score += notebookMetadata ? 1 : 8;
+                }
+
+                // See if the version is the same
+                if (interpreter && interpreter.version && spec && spec.name && nbMetadataLanguage === PYTHON_LANGUAGE) {
```

Comment:
I think this resolves #4941. I was about to take a look at this but figured that I should wait until the refactor. Looks like it should match on a python3 spec name to prefer a python3 interpreter over plain python.